File: Plan/src/main/java/com/djrapitops/plan/utilities/html/structure/SessionAccordion.java
package com.djrapitops.plan.utilities.html.structure;

import com.djrapitops.plan.api.PlanAPI;
import com.djrapitops.plan.data.container.PlayerKill;
import com.djrapitops.plan.data.container.Session;
import com.djrapitops.plan.data.store.keys.SessionKeys;
import com.djrapitops.plan.data.store.objects.DateHolder;
import com.djrapitops.plan.data.time.WorldTimes;
import com.djrapitops.plan.system.settings.Settings;
import com.djrapitops.plan.system.settings.theme.Theme;
import com.djrapitops.plan.system.settings.theme.ThemeVal;
import com.djrapitops.plan.utilities.comparators.DateHolderRecentComparator;
import com.djrapitops.plan.utilities.formatting.Formatter;
import com.djrapitops.plan.utilities.html.HtmlStructure;
import com.djrapitops.plan.utilities.html.graphs.Graphs;
import com.djrapitops.plan.utilities.html.graphs.pie.WorldPie;
import com.djrapitops.plan.utilities.html.icon.Icons;
import com.djrapitops.plan.utilities.html.tables.HtmlTables;

import java.util.*;
import java.util.function.Supplier;

/**
 * Utility for creating Session accordion html and javascript from Session objects.
 *
 * @author Rsl1122
 * @see com.djrapitops.plan.data.container.Session for object
 */
public class SessionAccordion extends AbstractAccordion {

    private final boolean forPlayer;
    private final List<Session> sessions;
    private final Supplier<Map<UUID, String>> serverNamesSupplier;
    private final Supplier<Map<UUID, String>> playerNamesSupplier;

    private final StringBuilder viewScript;
    private final boolean appendWorldPercentage;
    private int maxSessions;

    // TODO
    private Theme theme;
    private Graphs graphs;
    private HtmlTables tables;
    private Formatter<DateHolder> yearFormatter;
    private Formatter<Long> timeAmountFormatter;

    private SessionAccordion(boolean forPlayer, List<Session> sessions,
                             Supplier<Map<UUID, String>> serverNamesSupplier,
                             Supplier<Map<UUID, String>> playerNamesSupplier) {
        super("session_accordion");

        this.forPlayer = forPlayer;
        this.sessions = sessions;
        this.serverNamesSupplier = serverNamesSupplier;
        this.playerNamesSupplier = playerNamesSupplier;
        viewScript = new StringBuilder();

        maxSessions = Settings.MAX_SESSIONS.getNumber();
        if (maxSessions <= 0) {
            maxSessions = 50;
        }
        appendWorldPercentage = Settings.APPEND_WORLD_PERC.isTrue();

        addElements();
    }

    public static SessionAccordion forServer(List<Session> sessions, Supplier<Map<UUID, String>> serverNamesSupplier,
                                             Supplier<Map<UUID, String>> playerNamesSupplier) {
        return new SessionAccordion(false, sessions, serverNamesSupplier, playerNamesSupplier);
    }

    public static SessionAccordion forPlayer(List<Session> sessions, Supplier<Map<UUID, String>> serverNamesSupplier) {
        return new SessionAccordion(true, sessions, serverNamesSupplier, HashMap::new);
    }

    public String toViewScript() {
        return viewScript.toString();
    }

    private void addElements() {
        if (forPlayer) {
            addElementsForPlayer();
        } else {
            addElementsForServer();
        }
        // Requires refactoring of Session object to contain information about player and server
    }

    private void addElementsForServer() {
        Map<UUID, String> serverNames = serverNamesSupplier.get();
        Map<UUID, String> playerNames = playerNamesSupplier.get();
        sessions.sort(new DateHolderRecentComparator());

        int i = 0;
        for (Session session : sessions) {
            if (i >= maxSessions) {
                break;
            }

            String serverName = serverNames.getOrDefault(session.getValue(SessionKeys.SERVER_UUID).orElse(null), "Unknown");
            String playerName = playerNames.getOrDefault(session.getValue(SessionKeys.UUID).orElse(null), "Unknown");
            String sessionStart = yearFormatter.apply(session);

            WorldTimes worldTimes = session.getValue(SessionKeys.WORLD_TIMES).orElse(new WorldTimes(new HashMap<>()));
            WorldPie worldPie = graphs.pie().worldPie(worldTimes);
            String longestWorldPlayed = session.getValue(SessionKeys.LONGEST_WORLD_PLAYED).orElse("Unknown");

            boolean hasEnded = session.supports(SessionKeys.END);
            String sessionEnd = hasEnded ? yearFormatter.apply(() -> session.getUnsafe(SessionKeys.END)) : "Online";

            String length = (hasEnded ? "" : "(Online) ") + timeAmountFormatter.apply(session.getValue(SessionKeys.LENGTH).orElse(0L));
            String afk = (hasEnded ? "" : "(Inaccurate) ") + timeAmountFormatter.apply(session.getValue(SessionKeys.AFK_TIME).orElse(0L));

            int playerKillCount = session.getValue(SessionKeys.PLAYER_KILL_COUNT).orElse(0);
            int mobKillCount = session.getValue(SessionKeys.MOB_KILL_COUNT).orElse(0);
            int deaths = session.getValue(SessionKeys.DEATH_COUNT).orElse(0);

            String info = appendWorldPercentage
                    ? HtmlStructure.separateWithDots(sessionStart, longestWorldPlayed)
                    : sessionStart;
            String title = HtmlStructure.separateWithDots(playerName, info) + "<span class=\"pull-right\">" + length + "</span>";
            String htmlID = "" + session.getValue(SessionKeys.START).orElse(0L) + i;
            String worldHtmlID = "worldPieSession" + htmlID;

            String leftSide = new AccordionElementContentBuilder()
                    .addRowBold(Icons.SESSION_LENGTH, "Session Ended", sessionEnd)
                    .addRowBold(Icons.PLAYTIME, "Session Length", length)
                    .addRowBold(Icons.AFK_LENGTH, "AFK", afk)
                    .addRowBold(Icons.SERVER, "Server", serverName)
                    .addBreak()
                    .addRowBold(Icons.PLAYER_KILLS, "Player Kills", playerKillCount)
                    .addRowBold(Icons.MOB_KILLS, "Mob Kills", mobKillCount)
                    .addRowBold(Icons.DEATHS, "Deaths", deaths)
                    .toHtml();

            String rightSide = "<div id=\"" + worldHtmlID + "\" class=\"dashboard-donut-chart\"></div>" +
                    "<script>" +
                    "var " + worldHtmlID + "series = {name:'World Playtime',colorByPoint:true,data:" + worldPie.toHighChartsSeries() + "};" +
                    "var " + worldHtmlID + "gmseries = " + worldPie.toHighChartsDrilldown() + ";" +
                    "</script>";
            viewScript.append("worldPie(")
                    .append(worldHtmlID).append(", ")
                    .append(worldHtmlID).append("series, ")
                    .append(worldHtmlID).append("gmseries")
                    .append(");");

            List<PlayerKill> kills = session.getValue(SessionKeys.PLAYER_KILLS).orElse(new ArrayList<>());
            String leftBottom = tables.killsTable(kills, null).parseHtml();

            String link = PlanAPI.getInstance().getPlayerInspectPageLink(playerName);
            String rightBottom = "<a target=\"_blank\" href=\"" + link + "\"><button href=\"" + link +
                    "\" type=\"button\" class=\"pull-right btn bg-blue waves-effect\">" +
                    "<i class=\"material-icons\">person</i><span>INSPECT PAGE</span></button></a>";

            addElement(new AccordionElement(htmlID, title)
                    .setColor(theme.getValue(ThemeVal.PARSED_SESSION_ACCORDION))
                    .setLeftSide(leftSide + leftBottom)
                    .setRightSide(rightSide + rightBottom));
            i++;
        }
    }

    private void addElementsForPlayer() {
        Map<UUID, String> serverNames = serverNamesSupplier.get();
        sessions.sort(new DateHolderRecentComparator());

        int i = 0;
        for (Session session : new ArrayList<>(sessions)) {
            if (i >= maxSessions) {
                break;
            }

            String serverName = serverNames.getOrDefault(session.getValue(SessionKeys.SERVER_UUID).orElse(null), "Unknown");
            String sessionStart = yearFormatter.apply(session);

            WorldTimes worldTimes = session.getValue(SessionKeys.WORLD_TIMES).orElse(new WorldTimes(new HashMap<>()));
            WorldPie worldPie = graphs.pie().worldPie(worldTimes);
            String longestWorldPlayed = session.getValue(SessionKeys.LONGEST_WORLD_PLAYED).orElse("Unknown");

            boolean hasEnded = session.supports(SessionKeys.END);
            String sessionEnd = hasEnded ? yearFormatter.apply(() -> session.getValue(SessionKeys.END).orElse(0L)) : "Online";

            String length = (hasEnded ? "" : "(Online) ") + timeAmountFormatter.apply(session.getValue(SessionKeys.LENGTH).orElse(0L));
            String afk = (hasEnded ? "" : "(Inaccurate) ") + timeAmountFormatter.apply(session.getValue(SessionKeys.AFK_TIME).orElse(0L));

            int playerKillCount = session.getValue(SessionKeys.PLAYER_KILL_COUNT).orElse(0);
            int mobKillCount = session.getValue(SessionKeys.MOB_KILL_COUNT).orElse(0);
            int deaths = session.getValue(SessionKeys.DEATH_COUNT).orElse(0);

            String info = appendWorldPercentage
                    ? HtmlStructure.separateWithDots(sessionStart, longestWorldPlayed)
                    : sessionStart;
            String title = HtmlStructure.separateWithDots(serverName, info) + "<span class=\"pull-right\">" + length + "</span>";
            String htmlID = "" + session.getValue(SessionKeys.START).orElse(0L) + i;
            String worldHtmlID = "worldPieSession" + htmlID;

            String leftSide = new AccordionElementContentBuilder()
                    .addRowBold(Icons.SESSION_LENGTH, "Session Ended", sessionEnd)
                    .addRowBold(Icons.PLAYTIME, "Session Length", length)
                    .addRowBold(Icons.AFK_LENGTH, "AFK", afk)
                    .addRowBold(Icons.SERVER, "Server", serverName)
                    .addBreak()
                    .addRowBold(Icons.PLAYER_KILLS, "Player Kills", playerKillCount)
                    .addRowBold(Icons.MOB_KILLS, "Mob Kills", mobKillCount)
                    .addRowBold(Icons.DEATHS, "Deaths", deaths)
                    .toHtml();

            String rightSide = "<div id=\"" + worldHtmlID + "\" class=\"dashboard-donut-chart\"></div>" +
                    "<script>" +
                    "var " + worldHtmlID + "series = {name:'World Playtime',colorByPoint:true,data:" + worldPie.toHighChartsSeries() + "};" +
                    "var " + worldHtmlID + "gmseries = " + worldPie.toHighChartsDrilldown() + ";" +
                    "</script>";
            viewScript.append("worldPie(")
                    .append(worldHtmlID).append(", ")
                    .append(worldHtmlID).append("series, ")
                    .append(worldHtmlID).append("gmseries")
                    .append(");");

            List<PlayerKill> kills = session.getValue(SessionKeys.PLAYER_KILLS).orElse(new ArrayList<>());
            String leftBottom = tables.killsTable(kills, null).parseHtml();

            addElement(new AccordionElement(htmlID, title)
                    .setColor(theme.getValue(ThemeVal.PARSED_SESSION_ACCORDION))
                    .setLeftSide(leftSide + leftBottom)
                    .setRightSide(rightSide));

            i++;
        }
    }

}

File: Plan/common/src/main/resources/assets/plan/web/js/charts/sessionAccordion.js
function loadSessionAccordion(json, error) {
    if (error) {
        $('#sessions-overview').addClass('forbidden'); // TODO Figure out 403
        return;
    }

    sessionTable = $("#sessions-overview").find("#tableAccordion").find("tbody");

    if (!sessionTable) throw new Error("Session table not found with ids #session-overview #tableAccordion");

    var sessions = json.sessions;
    for (var i = 0; i < sessions.length; i++) {
        var session = sessions[i];
        var title = createAccordionTitle(i, session);
        var body = createAccordionBody(i, session);

        var worldSeries = {name: 'World Playtime', colorByPoint: true, data: session.world_series};
        var gmSeries = session.gm_series;

        sessionTable.append(title);
        sessionTable.append(body);

        worldPie("worldpie_" + i, worldSeries, gmSeries);
    }
}

function createAccordionTitle(i, session) {
    return '<tr aria-controls="session_t_' + i + '" aria-expanded="false" class="clickable collapsed bg-teal" data-target="#session_t_' + i + '" data-toggle="collapse"><td>'
        + session.name + '</td>'
        + '<td>' + session.start + '</td>'
        + '<td>' + session.length + '</td>'
        + '<td>' + session.most_used_world + '</td></tr>'
}

function createAccordionBody(i, session) {


    return '<tr class="collapse" data-parent="#tableAccordion" id="session_t_' + i + '">' +
        '<td colspan="4">' +
        '<div class="collapse row" data-parent="#tableAccordion" id="session_t_' + i + '">' +
        '<div class="col-xs-12  col-sm-12 col-md-6 col-lg-6">' +
        '<p><i class="col-teal far fa-fw fa-clock"></i> Session Ended<span class="float-right"><b>' + session.end + '</b></span></p>' +
        '<p><i class="col-green far fa-fw fa-clock"></i> Session Length<span class="float-right"><b>' + session.length + '</b></span></p>' +
        '<p><i class="col-grey far fa-fw fa-clock"></i> AFK<span class="float-right"><b>' + session.afk_time + '</b></span></p>' +
        '<p><i class="col-green fa fa-fw fa-server"></i> Server<span class="float-right"><b>' + session.server_name + '</b></span></p><br>' +
        '<p><i class="col-red fa fa-fw fa-crosshairs"></i> Player Kills<span class="float-right"><b>' + session.player_kills.length + '</b></span></p>' +
        '<p><i class="col-green fa fa-fw fa-crosshairs"></i> Mob Kills<span class="float-right"><b>' + session.mob_kills + '</b></span></p>' +
        '<p><i class=" fa fa-fw fa-skull"></i> Deaths<span class="float-right"><b>' + session.deaths + '</b></span></p><hr>' +
        createKillsTable(session.player_kills) +
        '</div><div class="col-xs-12  col-sm-12 col-md-6 col-lg-6">' +
        '<div id="worldpie_' + i + '" class="chart-pie"></div>' +
        '<a target="_blank" href="/player/' + session.player_name + '"><button type="button" class="float-right btn bg-blue waves-effect\"><i class="fa fa-user"></i><span> Player Page</span></button></a>' +
        '</div>' +
        '</div></td></tr>'
}

function createKillsTable(player_kills) {
    var table = '<table class="table table-striped scrollbar"><tbody>';

    if (player_kills.length === 0) {
        table += '<td>No Kills</td><td>-</td><td>-</td>'
    }

    for (var i = 0; i < player_kills.length; i++) {
        var kill = player_kills[i];
        table += '<td>' + kill.date + '</td>' +
            '<td>' + kill.killer + '<i class="fa fa-fw fa-angle-right col-red"></i>' + kill.victim + '</td>' +
            '<td>' + kill.weapon + '</td>'
    }

    table += '</tbody></table>';
    return table;
}